api-select.go

```diff
@@ -0,0 +1,627 @@
+	osreq.InputSerialization.CSV.FileHeaderInfo = opts.Input.FileHeaderInfo
+
+	// Only supports filed and record delimiter for now.
+	osreq.OutputSerialization.CSV.FieldDelimiter = opts.Output.FieldDelimiter
+	osreq.OutputSerialization.CSV.RecordDelimiter = opts.Output.RecordDelimiter
+
+	return osreq
+}
+
+// SelectObjectContent is a implementation of http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectSELECTContent.html AWS S3 API.
+func (c Client) SelectObjectContent(ctx context.Context, bucketName, objectName, expression string, opts SelectObjectOptions) (*Events, error) {
+	objInfo, err := c.statObject(ctx, bucketName, objectName, StatObjectOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	var compressionType = SelectCompressionNONE
+	if strings.Contains(objInfo.ContentType, "gzip") {
+		compressionType = SelectCompressionGZIP
+	} else if strings.Contains(objInfo.ContentType, "text/csv") && opts.Type == "" {
+		opts.Type = SelectObjectTypeCSV
+	} else if strings.Contains(objInfo.ContentType, "json") && opts.Type == "" {
+		opts.Type = SelectObjectTypeJSON
+	}
+
+	selectReqBytes, err := xml.Marshal(opts.ToObjectSelectRequest(expression, compressionType))
+
+	if err != nil {
+		return nil, err
+	}
+
+	urlValues := make(url.Values)
+	urlValues.Set("select", "")
+	urlValues.Set("select-type", "2")
+	// Execute POST on bucket/object.
+	resp, err := c.executeMethod(ctx, "POST", requestMetadata{
+		bucketName:       bucketName,
+		objectName:       objectName,
+		queryValues:      urlValues,
+		contentMD5Base64: sumMD5Base64(selectReqBytes),
+		contentSHA256Hex: sum256Hex(selectReqBytes),
+		contentBody:      bytes.NewReader(selectReqBytes),
+		contentLength:    int64(len(selectReqBytes)),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, httpRespToErrorResponse(resp, bucketName, "")
+	}
+	// Process needs to return a eventStream to the client
+	// Wrap this in a Go routine so as to allow the streaming
+	streamer := &Events{
+		Response: make(chan interface{}, 2),
+		lock:     make(chan bool, 2),
+	}
+	go func() {
+		processBinary(resp.Body, streamer)
+	}()
+	return streamer, nil
+}
+
+// processBinary is the main function that decodes the large byte array into
+// several events that are sent through the eventstream.
+func processBinary(message io.ReadCloser, streamer *Events) {
+	defer message.Close()
+	var partialRecordVal []byte
+	lenOfTrim := 0
+	for {
+		// select block for catching a user request for closing.
+		select {
+		// Logic is basically that the .Close() method will close the streamer.lock
+		// channel, Therefore if Ok is false(the channel is closed) then we can
+		// close our response channel and return. The .Close() method blocks on the
+		// streamer.Response channel being closed, so that it is guranteed that this
+		// method has exited since it is a defered close.
+		case _, ok := <-streamer.lock:
+			if !ok {
+				defer close(streamer.Response)
```

Comment:
no reason to defer before return you can just close and return.

Reply:
The reason for deferring, is that by deferring the closing of the channel until after the return, it is guaranteed that the processing thread will return before the the .Close() method returns. Otherwise without the defer, every once out of every 10 tests, I would get some weird behavior that given the timing of things, the .Close() returns before processing thread does. That might be okay, but I wasn't sure so to guarantee that the processing thread returns before the .Close() does, I used defer.